Replumbing: pre-populate the EVP namemap with commonly known names

levitte opened the pull request:
This adds ossl_namemap_empty(), to detect if a namemap is empty and
can thereby be pre-populated.

Related to #8967

levitte committed:
Replumbing: pre-populate the EVP namemap with commonly known names

This adds ossl_namemap_empty(), to detect if a namemap is empty and
can thereby be pre-populated.

Related to #8967

richsalz commented on crypto/evp/ossl_kdf_names.h:
```diff
@@ -0,0 +1,52 @@
+ * https://www.openssl.org/source/license.html
+ */
+
+/* ONLY included by evp_fetch.c */
```
Maybe .h is not the best name for what's really C code, especially if it should only be included ONCE?  .inc perhaps?

richsalz commented on crypto/evp/ossl_kdf_names.h:
```diff
@@ -0,0 +1,52 @@
+static const char * const id_pbkdf2_names[] =
+    { LN_id_pbkdf2, NULL };
+
+
```
any reason for double blank lines?

paulidale commented on crypto/core_namemap.c:
```diff
@@ -11,44 +11,133 @@
+
 typedef struct {
+    /* names and n_names are organized the same way as argv and argc */
+    const char **names;
```
Would make more sense to use a stack here?

Perhaps also for the body but that's slightly further from the stack behaviour.

levitte commented:
@paulidale, it would probably be more constructive to discuss the technical aspects of this PR here...

levitte commented:
This includes #9969, please review that one first

levitte commented:
I'm putting this on hold, because it should be reworked to work together with the changes in #8985, i.e. that one needs to be merged first

levitte commented:
(the hold label that I added was on a personal basis, not an OMC decision thing, so I removed it)

levitte committed:
fixup! Replumbing: pre-populate the EVP namemap with commonly known names

levitte commented:
So, this is essentially finished at this point.  Updating it was a breeze...

However, I wonder if this is still desirable.  We have lived without a pre-populated namemap for a while now, and I still see no signs of panic or horror with regards to that, so ...

mattcaswell commented:
I think this is a valuable thing to have. We want to make it as easy as possible for people to transition from one provider to another. At the moment all of our aliases are held in the default provider - but if you don't load that then you don't get the full list of aliases. Having them pre-populated in the namemap would mean we always have them.

levitte commented:
I put this in WIP because @mattcaswell and I realised that we need to keep the legacy method database (which is often wrongly called "object database") around, and I might as well use that to pre-populate the EVP namemap instead of having an intermediary array with the exact same information.

levitte committed:
fixup! Replumbing: pre-populate the EVP namemap with commonly known names

This changes how the initial EVP namemap is prepopulated.  Instead of
have a separate database of names, it simply pilfers the still existing
legacy method database.

levitte committed:
fixup! fixup! Replumbing: pre-populate the EVP namemap with commonly known names

levitte committed:
fixup! fixup! Replumbing: pre-populate the EVP namemap with commonly known names

levitte commented:
There's something very odd happening with the solution to pre-populate names from the legacy method db.

levitte committed:
squash! fixup! Replumbing: pre-populate the EVP namemap with commonly known names

This affected the way legacy NIDs are looked up in evp_cipher_from_dispatch()
and evp_md_from_dispatch().  Instead of trying to find the NID directly,
look up the legacy method structure and grab the NID from there.  The
reason is that NIDs can be aliases for other NIDs, which looked like
a clash even if wasn't really one.

levitte commented:
It seems I finally pulled through all corner cases (and there were a few).  If the CIs agree, I'll take this out of WIP.

levitte commented:
Not WIP any more

levitte commented:
Ping

paulidale commented on doc/internal/man3/ossl_namemap_new.pod:
```diff
@@ -40,6 +41,10 @@ new B<OSSL_NAMEMAP>.
 
+ossl_namemap_empty() checks if the given B<OSSL_NAMEMAP> is empty or
+not.
+This can be useful for pre-population purposes.
```
I'd omit this line, it's describing a use rather than the function.

levitte replied:
Okie

levitte commented:
Merged.

6a835fcfb10ba004498f9e39873db3d2b9011609 Replumbing: pre-populate the EVP namemap with commonly known names
